tools/verifier-helper/src/lib.rs

```diff
@@ -0,0 +1,148 @@
+use cosmwasm_std::{Addr, Decimal, DepsMut, Env, StdError, Uint128};
+use cw_utils::PaymentError;
+use lavs_apis::{
+    id::TaskId,
+    interfaces::voting::{TotalPowerResponse, VotingPowerResponse},
+    tasks::{TaskQueryMsg, TaskStatus, TaskStatusResponse},
+    verifier_simple::{OperatorVote, TaskMetadata},
+};
+
+use lavs_apis::interfaces::voting::QueryMsg as OperatorQueryMsg;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum VerifierError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("{0}")]
+    Payment(#[from] PaymentError),
+
+    #[error("Invalid percentage, must be between 1 and 100")]
+    InvalidPercentage,
+
+    #[error("Operator tried to vote twice: {0}")]
+    OperatorAlreadyVoted(String),
+
+    #[error("Task expired. Cannot vote on it")]
+    TaskExpired,
+
+    #[error("Task already completed. Cannot vote on it")]
+    TaskAlreadyCompleted,
+
+    #[error("Unauthorized")]
+    Unauthorized,
+    // Add any other custom errors you like here.
+    // Look at https://docs.rs/thiserror/1.0.21/thiserror/ for details.
+}
+
+/// Does all checks to ensure the voter is valid and has not voted yet.
+/// Also checks the task is valid and still open.
+/// Returns the metadata for the task (creating it if first voter), along with the voting power of this operator.
+///
+/// We do not want to error if an operator votes for a task that is already completed (due to race conditions).
+/// In that case, just return None and exit early rather than error.
+#[allow(clippy::too_many_arguments)]
+pub fn ensure_valid_vote(
+    mut deps: DepsMut,
+    env: &Env,
+    task_queue: &Addr,
+    task_id: TaskId,
+    operator: &Addr,
+    vote: Option<OperatorVote>,
+    metadata: &Option<TaskMetadata>,
+    fraction_percent: u32,
+    operators_addr: &Addr,
+) -> Result<Option<(TaskMetadata, Uint128)>, VerifierError> {
+    // Operator has not submitted a vote yet
+    if vote.is_some() {
```

Comment:
I would make this check in the calling code.
Great if it returns `VerifierError::OperatorAlreadyVoted`, but we do not want to constrain the vote type (it is different in oracle).

Happy to constrain the tasks stuff and the query for operator power. Maybe you can refactor this such that `load_or_initialize_metadata` is exposed (as a method on some Tasks newtype), and then `ensure_valid_vote` is converted to `get_operator_power` function.

This will simplify the flow and reduce the (very large) number of arguments.